### DISCUSSION TITLE
SwiftUI rollout plan?

### DISCUSSION THREAD

**Discussion by thedrick** [MobileNativeFoundation/discussions#116] [General]
Hey everyone, I don't think I've introduced myself yet - I'm Tyler Hedrick, a senior iOS engineer at Airbnb focused on UI frameworks and design systems. I'm curious how others are handling the introduction of SwiftUI. Our current strategy can be summarized as: we aren't using it until we feel that it's ready for production. Last year, we were thinking that iOS 14's version of SwiftUI might be ready enough for experimentation, but after all of the improvements announced in iOS 15 for SwiftUI, I'm not so sure.

Do other companies already have SwiftUI code running in production? How has that gone for you? If not, what are your plans to introduce it?

**Comment by gabriellanata** [comment]
Hello! Thanks for starting this thread. I'm an iOS engineer at Lyft in the Design Systems team, we also own anything related to UI frameworks in our apps. Sadly, we still support iOS 11 and won't be able to get to SwiftUI for a few years, so I can't answer the last questions. However, we have started preparing for it in a long term plan.

A few actions we took already:
1. We created a custom wrapper that makes UIKit syntax very similar to SwiftUI (Discussion related to that here: https://github.com/MobileNativeFoundation/discussions/discussions/53). Hopefully this will be open sourced soon.
2. Deprecated anything related to InterfaceBuilder.
3. We began preparing for reactive UI by embracing RxSwift across our codebase. Previously we were mostly using it for the networking side of things.

This has helped modernize our UI frameworks and should make translating to SwiftUI much easier in the future. Killing IB is a big win for readability of UI code and ability to code review.

When the time comes and we're ready to support SwiftUI we plan on doing it only at the view/viewController level for now. We have a very custom architecture and it would be extremely hard to adapt it for SwiftUI navigation, but it doesn't block us from creating the actual views using SwiftUI and wrapping them in a UIHostingController.

Hopefully this plan makes sense. We're not super familiar with any issues we could encounter while using a bunch of UIHostingControllers  so I'd love to know how well hybrid projects have worked out for others.

**Reply by thedrick** [reply]
This makes a lot of sense. I've also heard that SwiftUI navigation isn't quite up to production quality (though they might have fixed this in iOS 15). 

For the custom wrapper, we actually recently open sourced our version of this as well called [EpoxyLayoutGroups](https://github.com/airbnb/epoxy-ios/wiki/EpoxyLayoutGroups). We will likely stick with UIKit for a while longer especially given how much infrastructure we've built around it.

**Comment by mayankgandhi** [comment]
hey @thedrick! Thanks for starting this thread! I'm a big fan of SwiftUI's principles and declarative programming in general - I had previously worked in a 4 engineer team in shipping a short-form video app that was purely built-in SwiftUI and we noticed various bottlenecks which were optimized even in UIKit. There were multiple hiccups during the build process and tons of UIViewRepresentable wrappers over AVFoundation and CollectionView components. 

We found ourselves reinventing the wheel in various features and components like:

- Rendering Images and Videos in other viewcomponents
- Navigation throughout the app 
- Showing cards using UIPresentationController
- Using Collection views' optimized diffable APIs to show grids of multiple videos and images
- Building Camera, Video Player, and other AV foundation features
- Animating transitions and layout changes

On the flip side, we also noticed some improvements in the time taken in shipping features, No more Autolayout, a more stable state machine. Our Design team loved how quickly we could iterate through ideas and features. The Ability to write code for universal platforms with minimal tweaks- iOS, iPadOS, MacOS, TVOS, WatchOS is a huge deal especially once everything is running on ARM chips and SwiftUI is more prevalent.

I think that SwiftUI is at a very nascent stage and needs a lot of improvement and optimization from Apple before it can be used for production. It's a fantastic tool for building something simple for an MVP/Prototype but I think that iOS 17 would probably be the year when SwiftUI is adapted into

**Reply by thedrick** [reply]
super appreciate these details!

**Reply by Andrew-Chen-Wang** [reply]
iOS 17 has come out; would love to know your thoughts heading into this development cycle:)
